### PR TITLE
Implement Polynomial2D Operators

### LINKED ISSUE #1440: Implement The Rest Of The Operators For Polynomial2d
### Description of the task
Implement the equivalent set of operators for `Polynomial2d` to that which are implemented for `Polynomial1d`. Several operators, such as `==`, `+=`, and `+` are already implemented for `Polynomial2d`.

### Acceptance criteria

- [ ] Implement all operators for `Polynomial2d` that are currently implemented for `Polynomial1d` (_Note: if some don't make sense, then use ask a lead about maybe not implementing that particular function_)
- [ ] Test all added functionality.

### PR DESCRIPTION
**Pull request by kelvinkoon**
### Description

Implemented Polynomial2D operators that Polynomial1D operators were used for.
`*` and `*=` were not added as they were deemed unnecessary.

### Testing Done

Added tests to `polynomial2d_test.cpp`.

### Resolved Issues

Resolves #1440 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by MathewMacDougall — APPROVED** [review]
LGTM

File: src/software/new_geom/polynomial2d_test.cpp [hunk 1]
```diff
@@ -166,3 +177,16 @@ TEST(Polynomial2dTest, addition_assignment_operator)
...
+    const Polynomial2d p1(Polynomial1d({1, 2, 3}), Polynomial1d({1, 1, 1}));
+
+    Polynomial2d p2(Polynomial1d({6, 8, 10}), Polynomial1d({5, 2, 3}));
+
+    const Polynomial2d expected(Polynomial1d({5, 6, 7}), Polynomial1d({4, 1, 2}));
+
+    p2 -= p1;
+
+    EXPECT_EQ(expected, p2);
+}
```
Comment (garethellis0) [inline src/software/new_geom/polynomial2d_test.cpp]: Could we add two tests for the equality operator; equal and not equal?

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/software/new_geom/polynomial2d.cpp [hunk 2]
```diff
@@ -60,17 +60,27 @@ Polynomial1d Polynomial2d::getPolyY() const
...
 }
 
 Polynomial2d &operator+=(Polynomial2d &p1, const Polynomial2d &p2)
 {
     return p1 = p1 + p2;
 }
+
+Polynomial2d &operator-=(Polynomial2d &p1, const Polynomial2d &p2)
+{
+    return p1 = p1 - p2;
```
Comment (jonathanlew) [inline src/software/new_geom/polynomial2d.cpp]:
```suggestion
    return p1 - p2;
```

File: src/software/new_geom/polynomial2d.cpp [hunk 3]
```diff
@@ -60,17 +60,27 @@ Polynomial1d Polynomial2d::getPolyY() const
...
-    return Polynomial2d(p1.getPolyX() + p2.getPolyX(), p1.getPolyY() + p2.getPolyY());
+    return Polynomial2d(p1.getPolyX() - p2.getPolyX(), p1.getPolyY() - p2.getPolyY());
 }
 
 Polynomial2d &operator+=(Polynomial2d &p1, const Polynomial2d &p2)
 {
     return p1 = p1 + p2;
 }
+
+Polynomial2d &operator-=(Polynomial2d &p1, const Polynomial2d &p2)
```
Comment (jonathanlew) [inline src/software/new_geom/polynomial2d.cpp]:
```suggestion
Polynomial2d &operator-=(const Polynomial2d &p1, const Polynomial2d &p2)
```

File: src/software/new_geom/polynomial2d.cpp [hunk 4]
```diff
@@ -60,17 +60,27 @@ Polynomial1d Polynomial2d::getPolyY() const
...
 }
 
-Polynomial2d operator+(const Polynomial2d &p1, const Polynomial2d &p2)
+Polynomial2d operator-(const Polynomial2d &p1, const Polynomial2d &p2)
 {
-    return Polynomial2d(p1.getPolyX() + p2.getPolyX(), p1.getPolyY() + p2.getPolyY());
+    return Polynomial2d(p1.getPolyX() - p2.getPolyX(), p1.getPolyY() - p2.getPolyY());
 }
 
 Polynomial2d &operator+=(Polynomial2d &p1, const Polynomial2d &p2)
```
Comment (jonathanlew) [inline src/software/new_geom/polynomial2d.cpp]:
```suggestion
Polynomial2d &operator+=(const Polynomial2d &p1, const Polynomial2d &p2)
```

File: src/software/new_geom/polynomial2d.cpp [hunk 5]
```diff
@@ -60,17 +60,27 @@ Polynomial1d Polynomial2d::getPolyY() const
...
 {
     return p1 = p1 + p2;
```
Comment (jonathanlew) [inline src/software/new_geom/polynomial2d.cpp]:
```suggestion
    return p1 + p2;
```

File: src/software/new_geom/polynomial2d_test.cpp (hunk 1 above)
Comment (kelvinkoon) [reply src/software/new_geom/polynomial2d_test.cpp]: There are three tests  in `polynomial2d_test.cpp`.
`equality_operator_equal` tests the equal.
`equality_operator_x_poly_not_equal` and `equality_operator_y_poly_not_equal` test not equal.

File: src/software/new_geom/polynomial2d.h [hunk 6]
```diff
@@ -108,3 +106,25 @@ Polynomial2d operator+(const Polynomial2d &p1, const Polynomial2d &p2);
...
+
+/**
+ * Subtracts a Polynomial2d to another Polynomial2d
+ *
+ * @param p1 the Polynomial2d to subtract from.
+ * @param p2 the Polynomial2d to subtract.
+ *
+ * @return the new Polynomial2d
+ */
+Polynomial2d &operator-=(Polynomial2d &p1, const Polynomial2d &p2);
```
Comment (jonathanlew) [inline src/software/new_geom/polynomial2d.h]: Can we change all of these to return a `Polynomial2d` not a `Polynomial2d&`. In most cases, we're return a reference to a stack object

**Review by jonathanlew — CHANGES_REQUESTED** [review]
I don't think += and -= shouldn't modify inputs. Otherwise, just minor comments

**Comment by jonathanlew** [comment]
@kelvinkoon actually, try implementing it like this: https://stackoverflow.com/questions/4581961/c-how-to-overload-operator

EDIT: this is out of scope for this PR

**Review by jonathanlew — APPROVED** [review]

File: src/software/new_geom/polynomial2d_test.cpp (hunk 1 above)
Comment (garethellis0) [reply src/software/new_geom/polynomial2d_test.cpp]: My bad, thanks!

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**